### ISSUE TITLE
Fix vector tile overdrawing with overlapping tiles from different zoom levels

### ISSUE THREAD

**Issue by jfirebaugh** [mapbox/mapbox-gl-js#6768]
The tile loading logic in Mapbox GL JS attempts to load a parent tile in certain cases when a request for a child tile fails, e.g. the tile response is a 404, or a 200 with a content length of 0. This is intended to compensate for situations where tiles are not available at the current zoom level, but may be available at lower zoom levels, which could be overzoomed.

Unfortunately, this leads to overdraw in certain situations. I believe the primary situation (there may be others) involves:

* A vector tileset that does not cover the whole world, i.e. tiles exist only for a certain bounding box. Outside that bounding box, the Mapbox tile API will respond with 404s.
* Layers sourced from this tileset that use semi-transparent styling.
* A map viewport that intersects the bounds of the tileset.

Under these conditions, the map succeeds in loading one or more tiles at zoom `z` inside the bounds, but fails on one or more tiles outside the bounds. This triggers requests for tiles at `z-1` that succeed, returning tiles that overlap the child tiles. Both parent and child tiles are rendered, producing overdraw. (The fallback loading behavior was originally implemented for fully opaque raster tiles, and semi-transparent vector tiles was not taken into account.)

Multiple issues have been filed relating to this behavior:

* #4691, involving a transparent circle layer ([JSFiddle](https://jsfiddle.net/nlindblad/nucky1bm/)):
  ![mapbox-bug](https://cloud.githubusercontent.com/assets/164991/25857886/42fcb304-34d2-11e7-8ee4-dbf6e3f1636f.gif)
* #5829, similarly ([demo](https://www.mapbox.com/help/demos/heatmap-gljs-dds/demo-3.html)):
   ![circle-zoom](https://user-images.githubusercontent.com/8291663/33771091-635bce58-dbf5-11e7-958b-fef415a340d4.gif)
* #6386, similarly ([jsbin](http://jsbin.com/yobagowica/edit?html,output)
   ![v44dds](https://user-images.githubusercontent.com/17933661/37838375-1c89dad8-2e8e-11e8-8e4a-cfcc05a5e94e.gif)
* #6560, with a heatmap layer ([example](https://bl.ocks.org/j5kay/a50925604a76c892826e5d3a7db77ddd))
* #6754, with a fill-extrusion layer ([JSFiddle](https://jsfiddle.net/9ou7r119/5/))

IIRC, we originally limited the parent tile fallback behavior to raster sources. This issue probably dates to when we expanded that behavior to vector sources as well. I can think of a few ways we might fix it:

* Revert loading parent tiles for vector sources. We'd have to go back and look at the rationale for that change and see if there's a different way that doesn't trigger this bug. 
* Don't render parent tiles for vector sources when there's any successful child tile. Probably don't want this because parent tile rendering is important to avoid showing empty areas during pans and zooms.
* Adjust the parent tile loading behavior to be smarter around tileset bounds. Something like: don't load a parent tile when a child tile is near the bounds or when the parent tile intersects the bounds.
* Clip child tiles out of parent tiles during rendering. We did this for raster tiles in #5105.

@kkaefer, @mollymerp you're most familiar with tile loading / clipping behavior. What are your thoughts?

Making this high priority since it's affecting multiple customers. cc @lilykaiser

**Comment by springmeyer** [comment]
Another option would be, for vector sources via the Mapbox API, ignore the maxzoom defined in the tilejson and simply request tiles at a higher zoom. So, let the server overzoom: In this case the Mapbox API will do the 404 check internally and will clip out a new tile for you.

**Comment by springmeyer** [comment]
> I don't think the current issue is a problem with overzooming – its mainly an issue with sparse tilesets where some children of a parent tile have data and others do not, so the parent is retained and overdraws on the child tile.

@mollymerp I'm proposing not retaining/overdrawing but rather requesting tiles from the server and letting it overzoom. Or am I missing something?

**Comment by kkaefer** [comment]
I don't think we should rely on server-side overzooming, or ever request tiles that go beyond the TileJSON's `maxzoom`. Overzooming tiles is one of the hallmarks of GL and allows us to keep data usage low, and render high-res maps from offline tiles as well. Even if we did request server-side overzoomed tiles, we'd still have to add this logic to GL to support the use cases where those tiles haven't loaded yet, or will never load due to connectivity loss.

**Comment by mollymerp** [comment]
update here from some more investigations @ryanhamley and I did: 

- we tried the clipping via stencil buffer option, but unfortunately it won't work for unclipped layers (circle, heatmap, etc) because those features need to be allowed to cross tile boundaries. 
- turns out gl-native treats 404 as valid vector tiles with a "nocontent" flag, so the issue is not as prevalent because it only overdraws parent tiles for a frame or two until the children error out – we thought we could use this approach (#6768) but it prevents valid overzooming when you go past a tileset's maxzoom. (note that overzooming when tiles 404 at zoom level Z, but Z < source maxzoom does not work the same way in native because of this)
- @jfirebaugh suggested a check where we don't retain parent tiles if at least one child has failed && at least one child has errored (indicating a sparse tileset). this is the best solution we've tried so far, but it does not completely eliminate overdraw. In the case of a "grandparent" tile that only has at least one empty quadrant, if we zoom in to the grandchild ZL, at least one parent tile will have all siblings error (because they're empty) and so we'll go up and find the grandparent, which does not 404, so we retain it causing overdraw (my attempt at an illustration below). 

![img_5828](https://user-images.githubusercontent.com/2425307/41491552-968f2c90-70ae-11e8-9d08-7531d96fdd12.JPG)

I have two more ideas:
- revert to only retaining parents 1 ZL up from the ideal tile for all vector tilesets 
- add a 'sparse' option for vector tile sources, and when that option is `true` 
  -  don't retain parent if any of the current tile's siblings have successfully loaded (causes a flicker in between zoom levels)
  -  only retain parents 1 ZL up, and don't retain parent tiles if at least one child has failed && at least one child has errored

I'm open to other ideas if anyone has thoughts!

**Comment by peterqliu** [comment]
🚗  drive-by thought: in cases where only some child tiles fail, can we iterate through the features inside each successful child tile, identify the same feature in the parent tile, and cull those out of the parent, to avoid overdraw? 

Only thing is, I can't say how reliable the deduping step is, without some unique feature id.
